Tutorial: How to build an Insurance dapp in reach (#1023)
reach-sh/reach-lang: docs/src/tut/community-insurance/index.md

# {#tut} Community insurance

This tutorial walks through the creation of a Reach decentralized application with React (v16.8 - Hooks) as the frontend.

> It demonstrates how you can use reach together with react to build a real world community insurance application.

Reviewer: ```suggestion
It demonstrates how you can use Reach together with React to build a real-world community insurance application.
```

To follow along, you are expected to have already gone through [the most basic tutorial](##tut) and you have already
learnt how to develop and test with a command-line frontend. 
We highly encourage you to read about [the reach architecture](https://docs.reach.sh/rsh/#ref-programs) as well, 
as it's an excellent resource to help you understand the different modes/states of any Reach program.

## Assumptions
This tutorial assumes that you already
 * have [Docker](https://www.docker.com/get-started), and [Docker Compose](https://docs.docker.com/compose/install/) installed.
 * have node & npm installed on your machine. If not, first head over to the [official website](https://nodejs.org/en/download/) 
    and follow the installation steps.
 * have working knowledge of react hooks. You can checkout [the basics](https://reactjs.org/docs/hooks-overview.html#:~:text=Hooks%20are%20functions%20that%20let,if%20you'd%20like.) in case this is new to you.
 * installed Reach successfully by following the [installation process](##tut) from the basic tutorial. 

> Since this is a real world application, we want to style it professionally using [tailwind css](https://tailwindcss.com/). You don't have to be familiar with this framework to follow along though. 

Reviewer: ```suggestion
Since this is a real-world application, we want to style it professionally using [Tailwind CSS](https://tailwindcss.com/). You don't have to be familiar with this framework to follow along, though. 
```

We'll provide a step by step configuration process to enable this styling framework.

In case you want to first have some background knowledge on the insurance business, checkout [this pdf doc](https://ira.go.ug/cp/uploads/English%20Handbook%20final.pdf).

## Preparation
* Complete the most [basic tutorial](##tut), finish the installation of Reach and all related configuration such as Docker.

> * Initialize a react app (using [create-react-app](https://tailwindcss.com/docs/guides/create-react-app)) and configure tailwind css.

Reviewer: ```suggestion
* Initialize a React application (using [create-react-app](https://tailwindcss.com/docs/guides/create-react-app)) and configure Tailwind.css.
```

## Initialize the application
Go to the directory where you wish to create your project and open a new terminal from there.

Run these commands to initialise the project
```cmd
$ npx create-react-app community-insurance-dapp
$ cd community-insurance-dapp
```

> To Install Tailwind CSS,

Reviewer: ```suggestion
To install Tailwind CSS,
```

```cmd
$ npm install -D tailwindcss postcss autoprefixer
$ npx tailwindcss init -p
```

To Configure Tailwind CSS, open the file called `tailwind.config.js` at the root and paste in the following code.
```
load: /examples/insurance/tailwind.config.js
```
Open your `src/index.css` file and paste the code below, overwriting everything inside.
```
load: /examples/insurance/src/index.css
```

At the root of your application code, there a `package.json` file which contains all the configuration for react 

> and it is responsible for keeping track of all the dependances required to run this application. 

Reviewer: ```suggestion
and it is responsible for keeping track of all of the dependencies required to run this application.
```

> Open it and paste the code bellow, overwriting everything inside.

Reviewer: ```suggestion
Open it and paste the code below, overwriting everything inside.
```

```
load: /examples/insurance/package.json
```

> Now run the command bellow and wait for npm to finish installing all React dependances

Reviewer: ```suggestion
Now run the command below and wait for `npm` to finish installing all React dependencies.
```

```cmd
$ npm install
```
You have now finished preparing your react frontend app. What about the Reach backend ? 
Well, if you had successfully completed [the basic tutorial](##tut), by now you know that at the root of 

> the reach project there is a file named `index.rsh`, 

Reviewer: ```suggestion
the Reach project there is a file named `index.rsh`, 
```

> as well as the reach script file which allows you to run the command `./reach xxx`. 

Reviewer: ```suggestion
as well as the `reach` script, which allows you to run commands like `reach compile` and `reach run`. 
```

file resides inside the react project so that it can be imported at the top of react code. 
To achieve this, create a folder inside the src/ folder of the react app where the reach compilation output will go. 
While you can name it whatever you want, I suggest `reach-build`. So your reach compiled output should be place inside 
`community-insurance-dapp/src/reach-build`. By the way, you can direct the output to go there at the time of running your 
commpillation command. By default `./reach compile` command places the output in a `build` folder at the root of the reach app.
You can change the destination of the output by specifying the `-o` flag like so: 
```cmd
$ REACH_CONNECTOR_MODE="ALGO-browser" ./reach compile -o community-insurance-dapp/src/reach-build
```
The above command assumes that your `index.rsh` file is in the same directory with the `community-insurance-dapp` folder.
If that's not your case, you will have to adjest the command accordingly.

Preparation is done. Now lets start writing our Reach application code.

## Create your Reach program file for the community insurance dapp
```
load: /examples/insurance/index.rsh

> range: range: 1-60

Reviewer 2: It seems that the codebase has changed since the guide was drafted. Think you mean to load lines 1-54.
However, I'd suggest breaking it up more. Load `Insurer` and talk about their methods, then load `CommunityMember` and work through their code. 

A more detailed tutorial would have directories in differing stages of development and build the codebase over several steps, (as in the RPS tutorial) but you can introduce this as a guide if you'd prefer to introduce the entire codebase as a complete entity.

```
RECALL: Every reach program behaves like a state machine. It keeps changing "states", 
and in each state there are valid and invalid opearations. 
The states are called steps such as "step", "concensus -step" and "local-step".
In the code above, We have included comments to let you know when the program changes to a different state.

+ Lines 7 through 31 specify the participant who will deploy the contract, and acts as the Insurance provider. 
After deploying the contract, this "Insurer" participant will disconnect from it and leave it running for the 
community members to attach to it and interact with it to perform certain tasks such as paying monthly fees, 
raising insurance claims, approving claims, etc.
+ Lines 33 through 52 specify the Community Member API. An API is similar to a class of participants that can 
be doing many different things with the contract at the  same time by calling the specified API functions. 
Please read more about APIs and [see how to define them](https://docs.reach.sh/rsh/appinit/#rsh_API) and 
be sure to get familiar with their syntax.

> + Lines 55 is a configuration which has to be in a "step" part of the code. 

Reviewer 2: Is currently line 53

You can find out [what this config is all about here](https://docs.reach.sh/rsh/appinit/#p_7).

> + Lines 57 is where the "step" state ends. 

Reviewer 2: Line 54

Calling the init() terminates the initial step state and switches the program into a "concensus-step" state/mode.

> Now that we are in the concensus step mode, we can call any function that is valid for this step, such as `.only()`.

Reviewer 2: consensus

The next steps in our program require us to be in a concensus-step mode, but we are now in "step" mode after calling `commit()`.
+ On Line 71 we are calling `Insurer.publish()` in order to switch back to a concensus-step.

+ Line 75 defines a [Set](https://docs.reach.sh/rsh/consensus/#rsh_Set) of items called `registeredMembers`, which will keep a linear list of all community members knonwn to this 
insurance application. It only keeps their addresses, more info about the members is kept away from the blockchain (off-chain). 
+ Line 78 through 81 defines a [Map](https://docs.reach.sh/rsh/consensus/#rsh_Map) which will keep all open insurance claims. 
Once a claim is funded or has expired, it will be deleted from the list. 
+ Lines 85 through 92 define a Map to keep the owners of the open claims (ie, commucity members). 

Next, 
```
load: /examples/insurance/index.rsh
range: range: 94-251
```
In this part of the program we define all the operations that a community member can execute as they declared previously in the 
step part of the program where we specified the CommunityMember API.
Here we are coding the logic of the CommunityMember API functions one by one. 

> Notice that we are using the [parallelReduce(...)](https://docs.reach.sh/rsh/consensus/#rsh_parallelReduce) function, which is one of the [concensus transfers](https://docs.reach.sh/guide/ctransfers/) in Reach. 

Reviewer 2: Would suggest introducing as `parallelReduce`
Notice that we are using `parallelReduce`, which creates a race between `API` participants in a consensus step.

and the initial values of these values are 1 and 1 respectively as they are the arguments passed to the `parallelReduce()` on line 98.
The array `[1, 1]` input into parallelReduce([1, 1]) match the left side of the call to this function `[membersCount, claimsCount]`. `membersCount`=1 and `claimsCount`=1.
Later these values will be changed by call to API functions in case a certain API function returns an 
array with different values from the these initial ones. 
If you have visited the sysntax of the [parallelReduce()](https://docs.reach.sh/rsh/consensus/#rsh_parallelReduce) function, then by now you know that
the general structure of the call to the `parallelReduce` is like this: 
```js
const [var1,var2, ...] = parallelReduce([initialVal1, initialVal2, ...]).api(x, x, x, x).api(x, x, x, x).timeout(...);
```
where `x, x, x, x` in the `.api()` function referes to the arguments of the api() which are also functions.
Those functions are the real logic of our application in regard to what community members will be doing with the application, the way they will interact with it.

> We highly recommend reading more about the [syntax of the API.api()](https://docs.reach.sh/rsh/consensus/#rsh_parallelReduce.api) fucntion to get confirtable with 

Reviewer 2: comfortable ?

what each of the argument functions should do. Be sure to understand each of the argument function in detail because this is the most important part you need to be able to effectively 

> use API finctions in general. It's worth taking your time on each of these functions as it will not leave you at the same level of understanding of 

Reviewer 2: functions
